### PR TITLE
Transferring some functions here

### PR DESCRIPTION
**Pull request by rdmorin**
The two functions that load from GAMBLR.results need to exist in this repo so they work properly when run without a MAF input.

### CODE REVIEW

**Comment by rdmorin** [comment]
I realized this has some conflicts with Master that should be resolved. Stay tuned.

File: R/get_gambl_metadata.R [hunk 1]
```diff
@@ -548,10 +548,11 @@ get_gambl_metadata = function(seq_type_filter = "genome",
   
   # take `seq_type_priority` lines
```
Comment (rdmorin) [inline R/get_gambl_metadata.R:549]: @lkhilton I'm curious what you think about this change.

File: R/get_gene_expression.R [hunk 2]
```diff
@@ -1,57 +1,109 @@
+
+#' @title Check Gene Expression.
```
Comment (rdmorin) [inline R/get_gene_expression.R:2]: Documentation is incomplete but I think this will really simplify the process and make it more flexible for the users moving forward

File: R/get_gambl_metadata.R (hunk 1 above)
Comment (Kdreval) [reply R/get_gambl_metadata.R:549]: Was there an issue with metadata that this update is called to solve? I think should be fine, just curious whether there were any hidden bugs so we can potentially also implement something similar in data version of this function to keep outputs consistent

File: R/get_gene_expression.R [hunk 3]
```diff
@@ -77,218 +129,61 @@
...
-      if(!missing(hugo_symbols)){
-        gene_ids = hugo_symbols
-        gene_id_type = "Hugo_Symbol"
-        not_the_id_type = "ensembl_gene_id"
-      }else{
-        gene_ids = ensembl_gene_ids
-        gene_id_type = "ensembl_gene_id"
-        not_the_id_type = "Hugo_Symbol"
+  load_expression_by_samples = function(genes,samples,verbose){
+      split_dir = "/projects/rmorin/projects/gambl-repos/gambl-rmorin/results/icgc_dart/DESeq2-0.0_salmon-1.0/mrna--gambl-icgc-all/split"
```
Comment (Kdreval) [inline R/get_gene_expression.R:155]: I definitely missed this update and the new setup. The grep  was returning gene expression data almost immediately, although I usually tried with just a handful of genes so don't know if it behaves same way on the large number of genes. Imagining the for loop approach is way slower than current engine setup but this was probably tested/benchmarked. Is this directory going to be also supported in the remote setup on local computer?
Separately, this path I think should be added to config instead of here

File: R/get_gambl_metadata.R (hunk 1 above)
Comment (rdmorin) [reply R/get_gambl_metadata.R:549]: This was allowing some examples of exomes and genomes from the same biopsy to come through in separate rows. This could be seen as a bug because it didn't happen systematically. It would only happen when they were given different sample_id (and this isn't always the case). When the sample_id was consistent the collapse worked as expected. The examples I was trying to deal with were MCLs with a capture and genome library with different sample_id.

File: R/get_gene_expression.R
Comment (rdmorin) [inline R/get_gene_expression.R:1]: This is ready for review and some feedback on the new approach.

File: R/get_gambl_metadata.R (hunk 1 above)
Comment (Kdreval) [reply R/get_gambl_metadata.R:549]: Ah I see - thanks! Ok we don't have to worry about this in the GAMBLR.data version since there are no examples yet with both capture and genome :+1:

**Review by Kdreval — APPROVED** [review]
Thanks!

File: R/get_gene_expression.R (hunk 3 above)
Comment (rdmorin) [reply R/get_gene_expression.R]: Yes I need to move this to the config. I wanted to get confirmation from users of this function that the change was acceptable before I finalized such things. We would also need to automate the splitting of files by sample_id in order for this to continue working properly as new data is added to GAMBL. The performance improvements I was striving for here were mostly in relation to reducing the RAM footprint without slowing down the code. I think the code is actually faster now using grep but it would be interesting to compare. I could probably add another option to revert to the single file instead of scanning the per-sample files so we could do a head-to-head benchmark. What do you think?

Comment (Kdreval) [reply R/get_gene_expression.R]: The per-sample split would be easy to do with additional snakemake rule and small script. I can bechmark different setups on GSC, and it might be good to have the option to revert to the single file - and see how it benchmarks, too

File: R/get_gambl_metadata.R (hunk 1 above)
Comment (lkhilton) [reply R/get_gambl_metadata.R:549]: This is how it should have always been!!!! Can you please implement `group_by(biopsy_id, patient_id)` though - if there is ever a situation with NA or duplicate biopsy_ids (possible) you'd get unexpected results from just grouping by biopsy_id alone.

File: R/get_gambl_metadata.R [hunk 4]
```diff
@@ -563,10 +564,11 @@ get_gambl_metadata = function(seq_type_filter = "genome",
   # define which seq type is not priority
   seq_type_non_priotiry <- ifelse(seq_type_priority == "genome", "capture", "genome")
   # take only those lines that seq_type is not `seq_type_priority` (those were already solved)
-  all_meta <- dplyr::filter(all_meta, ! sample_id %in% samples_with_priority_seq_type)
+  all_meta <- dplyr::filter(all_meta, ! biopsy_id %in% samples_with_priority_seq_type)
```
Comment (lkhilton) [inline R/get_gambl_metadata.R:567]: Related to my comment above, it would be better to do:  
```
all_meta <- all_meta %>% 
    group_by(tissue_status, patient_id, biopsy_id) %>% 
    # When > 1 seq_type exists, take the specified priority seq_type
    slice_max(seq_type == seq_type_priority, n=1, with_ties = TRUE) %>% 
    # When still duplicates after this, take the line with the least NAs
    slice_min(na_count, n=1, with_ties = FALSE) %>% 
    ungroup()
```
This takes care of all of this in a single function, I think?

File: R/get_gene_expression.R [hunk 5]
```diff
@@ -1,294 +1,176 @@
...
-            for( prioritize_this_value in prioritize_rows_by[[based_column]] ){
-              k = multi_exp_split_i[,based_column] == prioritize_this_value
-              if( any(k) ){
-                multi_exp_split_i = multi_exp_split_i[k,]
-                break
-              } # else:
-              # if there is no row with this value to prioritize, keep the rows
-              # and try the lower-priority value of the next loop.
-            }
+        all_rows = fread(cmd = grep_cmd,col.names = c("ensembl_gene_id",
```
Comment (lkhilton) [inline R/get_gene_expression.R:108]: Can you make this `data.table::fread` so that this works if the user doesn't have `data.table` loaded in their environment?

File: R/get_gambl_metadata.R (hunk 4 above)
Comment (rdmorin) [reply R/get_gambl_metadata.R:567]: My understanding is that these were separated because we always want the mRNA row when it's requested and we only want to prioritize the two DNA seq_types. Would the code you suggested work that way? i.e would we get back mrna + genome (when both genome and capture available)?

File: R/check_gene_expression.R [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+  # collapse duplicates if any
+  if(any(expression_data_rows$multi_exp == 1) ){
+    if(verbose){
+      rows_total = filter(expression_data_rows,multi_exp==1) %>% nrow()
+      unique_redundant = filter(expression_data_rows,multi_exp==1) %>% pull(biopsy_id) %>% unique()
+      nid = length(unique_redundant)
+      print(paste(rows_total, "rows and ", nid, "unique biopsies are redundant, dropping redundancy"))
+      print(head(unique_redundant))
+    }
+    original = expression_data_rows %>% filter(!is.na(mrna_sample_id))
```
[the remark this answers is not in the document]
Comment (rdmorin) [reply R/check_gene_expression.R:68]: I'm not sure what preceding code you're referring to. Is it the duplicated section of code or all of it including mark_duplicates_biopsy? Most of this came directly from Vladimir's PR.

[the remark this answers is not in the document]
Comment (lkhilton) [reply R/check_gene_expression.R:68]: Yes I think we can eliminate all of lines 44 to 68 without affecting the output

File: R/get_gambl_metadata.R (hunk 4 above)
Comment (lkhilton) [reply R/get_gambl_metadata.R:567]: Hm, I see. This code block is hard to follow because it yo-yos between `all_meta` and `all_meta_res` and those objects aren't named for their contents. I'm really just looking for a solution that uses patient_id and biopsy_id to handle the normals with NA biopsy_id values, because I think that will break here

File: R/check_gene_expression.R (hunk 6 above)
[the remark this answers is not in the document]
Comment (rdmorin) [reply R/check_gene_expression.R:68]: I think Vladimir's intention with that code was to safeguard about unexpected duplicates that remain. I can try removing it to see if anything changes and eliminate it if it's redundant.

File: R/check_gene_expression.R [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+  
+  genome_meta =  dplyr::filter(these_samples_metadata,
+                               seq_type=="genome",
+                               patient_id %in% expression_data_rows$patient_id) %>%
+    select(sample_id,patient_id,biopsy_id) %>%
+    rename(c("genome_sample_id"="sample_id"))
+  expression_data_rows = left_join(expression_data_rows,genome_meta, by=c("patient_id","biopsy_id"))
+  
+  return(select(expression_data_rows,-multi_exp))
+}
```
Comment (rdmorin) [inline R/check_gene_expression.R:111]: The outputs below show that the `mark_duplicates_biopsy` function is accomplishing what it's supposed to (right?). 
The second output also shows that linking to the capture and genome sample doesn't cause any artificial duplication of rows. I've set this up as an optional output that is not returned by default. 
```
> check_gene_expression(verbose = T,show_linkages = F) %>% nrow()
[1] "starting with 2630 rows"                                                                                                                                                                               
[1] "2630 remain after collapsing on biopsy_id"
[1] "362 LINE 74: rows corresponding to  181 unique biopsies are redundant, dropping redundancy"
[1] "S01-4978"  "SU01-6860" "AS01-7426" "S01-12869" "S03-5298"  "US02-56"  
[1] "2449 remain after collapsing on biopsy_id"
[1] 2449
> check_gene_expression(verbose = T,show_linkages = T) %>% nrow()
[1] "starting with 2630 rows"                                                                                                                                                                               
[1] "2630 remain after collapsing on biopsy_id"
[1] "362 LINE 74: rows corresponding to  181 unique biopsies are redundant, dropping redundancy"
[1] "S01-4978"  "SU01-6860" "AS01-7426" "S01-12869" "S03-5298"  "US02-56"  
[1] "2449 remain after collapsing on biopsy_id"
[1] 2449                                
```

File: R/check_gene_expression.R [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+  
+  collapse_duplicates = TRUE
+
+  if(verbose){
+    nrows = nrow(expression_data_rows)
+    print(paste("starting with",nrows,"rows"))
+  }
+
+  
+  mark_duplicates_biopsy = function(ew){
```
Comment (lkhilton) [inline R/check_gene_expression.R:41]: This is fine and it works, but it'll never not bug me that it could be accomplished with 
```
expression_data_rows %>% 
    group_by(patient_id, biopsy_id) %>% 
    mutate(multi_exp = n() > 1)) %>% # Returns TRUE/FALSE but could be mutated to 1/0 if really needed
    ungroup()
````

File: R/check_gene_expression.R [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+    print(paste(now_row,"remain after collapsing on biopsy_id"))
+  }
+  
+  # collapse duplicates if any
+  if(any(expression_data_rows$multi_exp == 1) ){
+    if(verbose){
+      rows_total = filter(expression_data_rows,multi_exp==1) %>% nrow()
+      unique_redundant = filter(expression_data_rows,multi_exp==1) %>% pull(biopsy_id) %>% unique()
+      nid = length(unique_redundant)
+      print(paste(rows_total, "LINE 74: rows corresponding to ", nid, "unique biopsies are redundant, dropping redundancy"))
```
Comment (lkhilton) [inline R/check_gene_expression.R:66]: Why `LINE 74`? Also is redundancy different from duplicate?

File: R/check_gene_expression.R [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+      print(head(unique_redundant))
+    }
+    original = expression_data_rows %>% filter(!is.na(mrna_sample_id))
+    expression_data_rows = group_by(expression_data_rows, patient_id, biopsy_id) %>% 
+      slice_max(str_detect(protocol, "Ribo"), n=1, with_ties = TRUE) %>% 
+      slice_max(ffpe_or_frozen == "frozen", n=1, with_ties = TRUE) %>% 
+      ungroup()
+    
+    
+    expression_data_rows = mark_duplicates_biopsy(expression_data_rows)
```
Comment (lkhilton) [inline R/check_gene_expression.R:76]: Did you drop a check that confirmed that this output has no duplicates?

File: R/check_gene_expression.R [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+    
+    if(verbose){
+      now_row = nrow(expression_data_rows)
+      print(paste(now_row,"remain after collapsing on biopsy_id"))
+    }
+
+  }else{
```
Comment (lkhilton) [inline R/check_gene_expression.R:83]: This is the `else` for `if(any(expression_data_rows$multi_exp == 1) )`, right? I'm not following what happens if the condition is false... what happens here?

File: R/check_gene_expression.R (hunk 7 above)
Comment (lkhilton) [reply R/check_gene_expression.R:111]: I see that this is true, but now I am worried that `get_gambl_metadata` is not capable of returning metadata for some capture sample IDs...

File: R/check_gene_expression.R (hunk 9 above)
Comment (rdmorin) [reply R/check_gene_expression.R:66]: I used the word redundancy to refer to the full set of IDs which is 50% redundancy and 50% the rows we want to keep.

File: R/check_gene_expression.R (hunk 10 above)
Comment (rdmorin) [reply R/check_gene_expression.R:76]: I don't think so
